Utils/Elasticsearch/requests/dataset_aggregations.json

```diff
@@ -0,0 +1,138 @@
+/*
+  Aggregated summary:
+  Average not deleted dataset size, categorized by campaigns and output formats
+
+  There are many occasions when aggregations are required but search hits are not.
+  For these cases the hits can be ignored by setting size=0.
+*/
+GET prodsys_rucio_ami/task/_search
+{
+   "size": 0,
+   "aggs": {
+     "campaign": {
+       "terms": {
+         "field": "campaign.keyword"
+       },
+        "aggs": {
+         "datasets": {
+           "children": {
+             "type": "output_dataset"
+           },
+           "aggs": {
+             "data_format": {
+               "terms": {
+                 "field": "data_format"
+               },
+                "aggs": {
+                 "bytes": {
+                   "filter": {
+                     "term": {
+                       "deleted": false
+                     }
+                   },
+                  "aggs": {
+                     "avg_bytes": {
+                       "avg": {
+                          "field": "bytes"
+                        }
+                     }
+                   }
+                 }
+               }
+             }
+           }
+         }
+       }
+     }
+   }
+}
+
+
+
+/*
+  Real example:
+  Average dataset size for tasks, filtered by query string and
+  with data format DAOD
+  Search by tasks
+*/
+
+GET prodsys_rucio_ami/task/_search
+{
+   "size": 0,
+   "query": {
+      "query_string": {
+           "query": "\"MC15a\" AND \"higgs\""
+         }
+   },
+   "aggs": {
+     "datasets": {
+       "children": {
+         "type": "output_dataset"
+       },
+       "aggs": {
+         "bytes": {
+            "filter": {
+              "bool": {
+                "must": [
+                  {"term": {"deleted": false }},
+                  {"term": {"data_format": "DAOD" }}
+                  ]
+              }
+             },
+           "aggs": {
+             "avg_bytes": {
+               "avg": {
+                "field": "bytes"
+              }
+             }
+           }
+         }
+       }
+     }
+   }
+ }
+
+/*
+  Real example:
+  Average dataset size
+  Search by datasets
+*/
+GET prodsys_rucio_ami/output_dataset/_search
```

Comment:
Should this query return average dataset size for all datasets?
I ask because there is a `should` clause:
```
"should": [
  {"term": {"data_format.keyword": "AOD"}},
  {"term": {"data_format.keyword": "DAOD"}}
]
```
If we wanted to get average size for AOD and DAOD datasets, then it does not work (if I specify `"size": 10"` I see datasets with `"data_format" : ["NTUP_PILEUP", "NTUP"]` as well). If we didn`t -- then this clause should be removed.

And, by the way, there\`s no `data_format.keyword` subfield in the current version of mapping (in `master`); `data_format` itself is of type `keyword`.

Reply:
Thank you! It was fixed in commit #ac069fc

Reply:
Thank you!

However, I have a couple of comments about this commit.

1. ac069fc description says:
```
ES/agg datasets query fixed

- 'Should' (OR) clause was moved inside the 'must' (AND) clause
- .keyword suffix removed from 'data_format'
```

How does this change correspond to it?
```
-            "query":
-              {"term": {"project.keyword": "mc15_13TeV"}}
+            "query":  {"term": {"project.keyword": "mc15_13TeV"}}
```

2. You change both formatting and query structure (and in the resulting query the indentation is a bit wierd), so it is very hard to get the essence of the change.

Look at 743fd7d -- it is perfectly clear, what and where was moved, as the query was uniformly formatted originally and follows the same style in the new version.

I would strongly recommend you to unify the formatting of all queries the way you take most readable (say, it is fair enough to place `"query":  {"term": {"key": "value"}}` on one line instead of 5 lines, that would be produced if we did it strictly); and then follow this formatting whatever changes you do.
Maybe even do it in the first commit and update next ones to follow the chosen variant.

I know it will take time and will be painfull. But isn`t it painful to me -- stumbling over this issue again and again, despite asking you to fix it? :(

3. I would also suggest more detailed description, that would explain why and where it was moved, something like:
```
We need the query to behave like "A and B and (C or D)"

Was: 
  "bool": {
    "must": [ A, B ],
    "should": [ C, D ],
  }

Here "should" clause affects only the order of the query results (<link to the ES docs>).
To make its conditions required, it must be moved inside the "must" clause:
  "bool": {
    "must": [
      A, B,
      "bool": { "should": [ C, D ] }
    ]
  }
```

It takes a bit more time to make this description, but it woulf significantly simplify the grasp of the commit.
I bet you yourself wouldn`t recall this logic a month from now, at least not without looking around the commit. Why make others to speculate? ;)

Reply:
[1] Should I fix it using rebase? 
[2] formatting was fixed in the last commit
[3] message about MUST&SHOULD usage was added to commit message - 165d008

Reply:
[1] Yes, sooner or later it seems to require a rebase. I check the queries as a whole all the time (as for now I am not good enough in ES query logic and can\`t juts "update" it in memory after some changes :) ), so if you feel like rewriting the history -- go ahead any time you want :)

[2] Thank you. Thought some issues are left... I will add comments to the code a bit later.

[3] It was not exactly about this commit. In 165d008 we did have 
```
  "bool": {
    "must": [ A, B ],
    "should": [ C ],
  }
```
originally, but new version is 
```
  "bool": {
    "must": [ A, B, C ]
  }
```
, right?